sql/tests: TestRandomSyntaxGeneration failed

sql/tests.TestRandomSyntaxGeneration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/21298692?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/21298692?buildTab=artifacts#/) on master @ [f1b136a030d439e5dc89a27cb4de08045dca8e77](https://github.com/cockroachdb/cockroach/commits/f1b136a030d439e5dc89a27cb4de08045dca8e77):


```
    rsg_test.go:992: 3m35s of 5m0s: 223011 executions, 11330 successful
    rsg_test.go:992: 3m40s of 5m0s: 227031 executions, 11497 successful
    rsg_test.go:992: 3m45s of 5m0s: 232169 executions, 11690 successful
    rsg_test.go:992: 3m50s of 5m0s: 237518 executions, 11878 successful
    rsg_test.go:992: 3m55s of 5m0s: 245505 executions, 12188 successful
    rsg_test.go:992: 4m0s of 5m0s: 250320 executions, 12392 successful
    rsg_test.go:992: 4m5s of 5m0s: 252437 executions, 12462 successful
    rsg_test.go:992: 4m10s of 5m0s: 253572 executions, 12510 successful
    rsg_test.go:992: 4m15s of 5m0s: 255879 executions, 12582 successful
    rsg_test.go:992: 4m20s of 5m0s: 258313 executions, 12680 successful
panic: /Table/3/1/138/2/1: failed to unmarshal descriptor {[209 240 220 87 3 26 144 1 8 104 16 133 1 26 7 95 102 97 109 105 108 121 32 138 1 40 1 58 57 8 15 16 0 24 0 42 0 80 169 142 6 90 41 8 15 16 0 24 0 42 0 80 169 142 6 90 19 8 6 16 0 24 0 42 0 80 162 9 96 0 106 4 8 0 16 0 96 0 122 4 16 170 142 6 96 0 64 0 72 1 82 0 96 137 1 104 0 114 48 10 11 10 5 97 100 109 105 110 16 2 24 2 10 13 10 6 112 117 98 108 105 99 16 128 4 24 0 10 10 10 4 114 111 111 116 16 2 24 2 18 4 114 111 111 116 24 3 122 0 160 1 0] 1775366109.921448016,0}
	(1) attached stack trace
	  -- stack trace:
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:783
	  | [...repeated from below...]
	Wraps: (2) attached stack trace
	  -- stack trace:
	  | github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic
	  | 	pkg/util/log/logcrash/crash_reporting.go:432
	  | github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigsqlwatcher.(*SQLWatcher).watchForDescriptorUpdates.func1
	  | 	pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go:215
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).processEvent
	  | 	pkg/kv/kvclient/rangefeed/rangefeed.go:453
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).processEvents
	  | 	pkg/kv/kvclient/rangefeed/rangefeed.go:439
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).run.func3
	  | 	pkg/kv/kvclient/rangefeed/rangefeed.go:393
	  | github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait.Group.GoCtx.func1
	  | 	pkg/util/ctxgroup/ctxgroup.go:200
	  | golang.org/x/sync/errgroup.(*Group).Go.func1
	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
	  | runtime.goexit
	  | 	src/runtime/asm_amd64.s:1693
	Wraps: (3) /Table/3/1/138/2/1: failed to unmarshal descriptor {[209 240 220 87 3 26 144 1 8 104 16 133 1 26 7 95 102 97 109 105 108 121 32 138 1 40 1 58 57 8 15 16 0 24 0 42 0 80 169 142 6 90 41 8 15 16 0 24 0 42 0 80 169 142 6 90 19 8 6 16 0 24 0 42 0 80 162 9 96 0 106 4 8 0 16 0 96 0 122 4 16 170 142 6 96 0 64 0 72 1 82 0 96 137 1 104 0 114 48 10 11 10 5 97 100 109 105 110 16 2 24 2 10 13 10 6 112 117 98 108 105 99 16 128 4 24 0 10 10 10 4 114 111 111 116 16 2 24 2 18 4 114 111 111 116 24 3 122 0 160 1 0] 1775366109.921448016,0}
	Error types: (1) *withstack.withStack (2) *withstack.withStack (3) *errutil.leafError

goroutine 2040 [running, "job":"AUTO SPAN CONFIG RECONCILIATION id=1164232118191226881", "n":"1"]:
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.Wait({0xc0115fe380?, {0x87fb698?, 0xc0115eb1d0?}, 0xc0115e29f0?})
	pkg/util/ctxgroup/ctxgroup.go:151 +0xbb
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait({0x87fb6d0?, 0xc0115ed800?}, {0xc011375cb0, 0x2, 0x6bf50c0?})
	pkg/util/ctxgroup/ctxgroup.go:224 +0x11d
github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).run(0xc0114f4000, {0x87fb6d0, 0xc0115ed800}, {0x886ab60, 0xc01149d1d0}, 0x0)
	pkg/kv/kvclient/rangefeed/rangefeed.go:396 +0xabd
github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).start.func1({0x87fb698?, 0xc01149d220?})
	pkg/kv/kvclient/rangefeed/rangefeed.go:256 +0x2e9
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func1({0x87fb698, 0xc01149d220})
	pkg/util/stop/stopper.go:447 +0x8a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 1425
	pkg/util/stop/stopper.go:445 +0x125
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #132558 sql/tests: TestRandomSyntaxGeneration failed [CREATE TENANT statement is not valid after parsing then formatting] [C-test-failure O-robot P-3 T-db-server branch-release-24.1.6-rc]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxGeneration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-62525

sql/tests.TestRandomSyntaxGeneration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/21312874?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/21312874?buildTab=artifacts#/) on master @ [27ba7199b3f7f933e835c8e946263c5084ab1b86](https://github.com/cockroachdb/cockroach/commits/27ba7199b3f7f933e835c8e946263c5084ab1b86):

Fatal error:

```
panic: /Table/3/1/135/2/1: failed to unmarshal descriptor {[68 145 234 198 3 26 139 1 8 104 16 29 26 9 95 103 114 111 117 112 105 110 103 32 135 1 40 1 58 51 8 15 16 0 24 0 42 0 80 166 142 6 90 35 8 15 16 0 24 0 42 0 80 166 142 6 90 13 8 2 16 32 24 0 42 0 80 188 5 96 0 96 0 122 4 16 167 142 6 96 0 64 0 72 1 82 0 96 134 1 104 0 114 48 10 11 10 5 97 100 109 105 110 16 2 24 2 10 13 10 6 112 117 98 108 105 99 16 128 4 24 0 10 10 10 4 114 111 111 116 16 2 24 2 18 4 114 111 111 116 24 3 122 0 160 1 0] 1776230186.390612094,0}
	(1) attached stack trace
	  -- stack trace:
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:783
	  | [...repeated from below...]
	Wraps: (2) attached stack trace
	  -- stack trace:
	  | github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic
	  | 	pkg/util/log/logcrash/crash_reporting.go:432
	  | github.com/cockroachdb/cockroach/pkg/spanconfig/spanconfigsqlwatcher.(*SQLWatcher).watchForDescriptorUpdates.func1
	  | 	pkg/spanconfig/spanconfigsqlwatcher/sqlwatcher.go:215
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).processEvent
	  | 	pkg/kv/kvclient/rangefeed/rangefeed.go:453
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).processEvents
	  | 	pkg/kv/kvclient/rangefeed/rangefeed.go:439
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).run.func3
	  | 	pkg/kv/kvclient/rangefeed/rangefeed.go:393
	  | github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait.Group.GoCtx.func1
	  | 	pkg/util/ctxgroup/ctxgroup.go:200
	  | golang.org/x/sync/errgroup.(*Group).Go.func1
	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
	  | runtime.goexit
	  | 	src/runtime/asm_amd64.s:1693
	Wraps: (3) /Table/3/1/135/2/1: failed to unmarshal descriptor {[68 145 234 198 3 26 139 1 8 104 16 29 26 9 95 103 114 111 117 112 105 110 103 32 135 1 40 1 58 51 8 15 16 0 24 0 42 0 80 166 142 6 90 35 8 15 16 0 24 0 42 0 80 166 142 6 90 13 8 2 16 32 24 0 42 0 80 188 5 96 0 96 0 122 4 16 167 142 6 96 0 64 0 72 1 82 0 96 134 1 104 0 114 48 10 11 10 5 97 100 109 105 110 16 2 24 2 10 13 10 6 112 117 98 108 105 99 16 128 4 24 0 10 10 10 4 114 111 111 116 16 2 24 2 18 4 114 111 111 116 24 3 122 0 160 1 0] 1776230186.390612094,0}
	Error types: (1) *withstack.withStack (2) *withstack.withStack (3) *errutil.leafError

```
Stack: 

```
goroutine 2137 [running, "job":"AUTO SPAN CONFIG RECONCILIATION id=1167063761789583361", "n":"1"]:
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.Wait({0xc01158c3c0?, {0x88aa0d8?, 0xc0115c8050?}, 0xc0114fe378?})
	pkg/util/ctxgroup/ctxgroup.go:151 +0xbb
github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait({0x88aa110?, 0xc0115943f0?}, {0xc0115dbcb0, 0x2, 0x6c42660?})
	pkg/util/ctxgroup/ctxgroup.go:224 +0x11d
github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).run(0xc01107da40, {0x88aa110, 0xc0115943f0}, {0x8919520, 0xc0114df810}, 0x0)
	pkg/kv/kvclient/rangefeed/rangefeed.go:396 +0xabd
github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).start.func1({0x88aa0d8?, 0xc0114df860?})
	pkg/kv/kvclient/rangefeed/rangefeed.go:256 +0x2e9
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func1({0x88aa0d8, 0xc0114df860})
	pkg/util/stop/stopper.go:447 +0x8a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 1634
	pkg/util/stop/stopper.go:445 +0x125
panic: /Table/3/1/135/2/1: unable to unmarshal descriptor {[68 145 234 198 3 26 139 1 8 104 16 29 26 9 95 103 114 111 117 112 105 110 103 32 135 1 40 1 58 51 8 15 16 0 24 0 42 0 80 166 142 6 90 35 8 15 16 0 24 0 42 0 80 166 142 6 90 13 8 2 16 32 24 0 42 0 80 188 5 96 0 96 0 122 4 16 167 142 6 96 0 64 0 72 1 82 0 96 134 1 104 0 114 48 10 11 10 5 97 100 109 105 110 16 2 24 2 10 13 10 6 112 117 98 108 105 99 16 128 4 24 0 10 10 10 4 114 111 111 116 16 2 24 2 18 4 114 111 111 116 24 3 122 0 160 1 0] 1776230186.390612094,0}
	(1) attached stack trace
	  -- stack trace:
	  | runtime.gopanic
	  | 	GOROOT/src/runtime/panic.go:783
	  | [...repeated from below...]
	Wraps: (2) attached stack trace
	  -- stack trace:
	  | github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic
	  | 	pkg/util/log/logcrash/crash_reporting.go:432
	  | github.com/cockroachdb/cockroach/pkg/sql/catalog/lease.(*Manager).watchForUpdates.func1
	  | 	pkg/sql/catalog/lease/lease.go:3141
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).processEvent
	  | 	pkg/kv/kvclient/rangefeed/rangefeed.go:453
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).processEvents
	  | 	pkg/kv/kvclient/rangefeed/rangefeed.go:439
	  | github.com/cockroachdb/cockroach/pkg/kv/kvclient/rangefeed.(*RangeFeed).run.func3
	  | 	pkg/kv/kvclient/rangefeed/rangefeed.go:393
	  | github.com/cockroachdb/cockroach/pkg/util/ctxgroup.GoAndWait.Group.GoCtx.func1
	  | 	pkg/util/ctxgroup/ctxgroup.go:200
	  | golang.org/x/sync/errgroup.(*Group).Go.func1
	  | 	external/org_golang_x_sync/errgroup/errgroup.go:93
	  | runtime.goexit
	  | 	src/runtime/asm_amd64.s:1693
	Wraps: (3) /Table/3/1/135/2/1: unable to unmarshal descriptor {[68 145 234 198 3 26 139 1 8 104 16 29 26 9 95 103 114 111 117 112 105 110 103 32 135 1 40 1 58 51 8 15 16 0 24 0 42 0 80 166 142 6 90 35 8 15 16 0 24 0 42 0 80 166 142 6 90 13 8 2 16 32 24 0 42 0 80 188 5 96 0 96 0 122 4 16 167 142 6 96 0 64 0 72 1 82 0 96 134 1 104 0 114 48 10 11 10 5 97 100 109 105 110 16 2 24 2 10 13 10 6 112 117 98 108 105 99 16 128 4 24 0 10 10 10 4 114 111 111 116 16 2 24 2 18 4 114 111 111 116 24 3 122 0 160 1 0] 1776230186.390612094,0}
	Error types: (1) *withstack.withStack (2) *withstack.withStack (3) *errutil.leafError
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestRandomSyntaxGeneration
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/d437d2c847dfedbc4972f231c3331c8e/logTestRandomSyntaxGeneration2610087653
    test_log_scope.go:82: use -show-logs to present logs inline
    rsg_test.go:992: 5s of 5m0s: 9472 executions, 1092 successful
    rsg_test.go:992: 10s of 5m0s: 17924 executions, 1899 successful
    rsg_test.go:992: 15s of 5m0s: 25062 executions, 2512 successful
    rsg_test.go:992: 20s of 5m0s: 29703 executions, 2898 successful
    rsg_test.go:992: 25s of 5m0s: 32453 executions, 3112 successful
    rsg_test.go:992: 30s of 5m0s: 32453 executions, 3112 successful
    rsg_test.go:992: 35s of 5m0s: 32950 executions, 3151 successful
    rsg_test.go:992: 40s of 5m0s: 36004 executions, 3381 successful
    rsg_test.go:992: 45s of 5m0s: 37584 executions, 3504 successful
    rsg_test.go:992: 50s of 5m0s: 39434 executions, 3644 successful
    rsg_test.go:992: 55s of 5m0s: 41088 executions, 3764 successful
    rsg_test.go:992: 1m0s of 5m0s: 43779 executions, 3958 successful
    rsg_test.go:992: 1m5s of 5m0s: 44846 executions, 4039 successful
    rsg_test.go:992: 1m10s of 5m0s: 46259 executions, 4129 successful
    rsg_test.go:992: 1m15s of 5m0s: 49804 executions, 4380 successful
    rsg_test.go:992: 1m20s of 5m0s: 55843 executions, 4766 successful
    rsg_test.go:992: 1m25s of 5m0s: 62430 executions, 5220 successful
    rsg_test.go:992: 1m30s of 5m0s: 68222 executions, 5591 successful
    rsg_test.go:992: 1m35s of 5m0s: 75037 executions, 6014 successful
    rsg_test.go:992: 1m40s of 5m0s: 82492 executions, 6477 successful
    rsg_test.go:992: 1m45s of 5m0s: 88764 executions, 6858 successful
    rsg_test.go:992: 1m50s of 5m0s: 94742 executions, 7123 successful
    rsg_test.go:992: 1m55s of 5m0s: 100324 executions, 7397 successful
    rsg_test.go:992: 2m0s of 5m0s: 105876 executions, 7669 successful
    rsg_test.go:992: 2m5s of 5m0s: 108636 executions, 7794 successful
    rsg_test.go:992: 2m10s of 5m0s: 108978 executions, 7811 successful
    rsg_test.go:992: 2m15s of 5m0s: 111078 executions, 7910 successful
    rsg_test.go:992: 2m20s of 5m0s: 111078 executions, 7910 successful
    rsg_test.go:992: 2m25s of 5m0s: 112193 executions, 7972 successful
    rsg_test.go:992: 2m30s of 5m0s: 115171 executions, 8099 successful
    rsg_test.go:992: 2m35s of 5m0s: 121323 executions, 8380 successful
    rsg_test.go:992: 2m40s of 5m0s: 123664 executions, 8478 successful
    rsg_test.go:992: 2m45s of 5m0s: 128265 executions, 8686 successful
    rsg_test.go:992: 2m50s of 5m0s: 136432 executions, 9068 successful
    rsg_test.go:992: 2m55s of 5m0s: 143045 executions, 9377 successful
    rsg_test.go:992: 3m0s of 5m0s: 149768 executions, 9645 successful
    rsg_test.go:992: 3m5s of 5m0s: 152380 executions, 9755 successful
    rsg_test.go:992: 3m10s of 5m0s: 156734 executions, 9969 successful
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #132558 sql/tests: TestRandomSyntaxGeneration failed [CREATE TENANT statement is not valid after parsing then formatting] [C-test-failure O-robot P-3 T-db-server branch-release-24.1.6-rc]
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxGeneration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/create_type.go
+++ b/pkg/sql/create_type.go
@@ -566,6 +566,10 @@ func (p *planner) createDomainWithID(
 	if baseType.TypeMeta.DomainData != nil {
 		return unimplemented.NewWithIssue(165347, "domain of domain is not yet supported")
 	}
+	// Reject domain of arrays and tuples types as well.
+	if baseType.Family() == types.ArrayFamily || baseType.Family() == types.TupleFamily {
+		return unimplemented.NewWithIssue(32552, "domain of arrays and tuples is not yet supported")
+	}
 
 	// Build CHECK constraints. Each expression is validated by substituting
 	// VALUE with a typed null of the base type and type-checking as boolean.

--- a/pkg/sql/logictest/testdata/logic_test/domain
+++ b/pkg/sql/logictest/testdata/logic_test/domain
@@ -422,6 +422,21 @@ CREATE DOMAIN d_nested AS d_base_for_nested
 
 subtest end
 
+subtest domain_of_array_and_tuple
+
+# Test that CREATE DOMAIN with an array base type is rejected (#32552).
+statement error pgcode 0A000 unimplemented: domain of arrays and tuples is not yet supported
+CREATE DOMAIN d_array AS INT[]
+
+# Test that CREATE DOMAIN with a tuple base type is rejected (#32552).
+statement ok
+CREATE TYPE my_tuple AS (a INT, b INT)
+
+statement error pgcode 0A000 unimplemented: domain of arrays and tuples is not yet supported
+CREATE DOMAIN d_tuple AS my_tuple
+
+subtest end
+
 subtest domain_drop_syntax
 
 # Test DROP DOMAIN syntax variants.

--- a/pkg/sql/types/types.go
+++ b/pkg/sql/types/types.go
@@ -1442,6 +1442,9 @@ func MakeDomain(baseType *T, typeOID, arrayTypeOID oid.Oid) *T {
 	it.UDTMetadata = &PersistentUserDefinedTypeMetadata{
 		ArrayTypeOID: arrayTypeOID,
 	}
+	if baseType.Family() == ArrayFamily || baseType.Family() == TupleFamily {
+		panic(errors.AssertionFailedf("cannot make domain of array or tuple"))
+	}
 	return &T{InternalType: it}
 }
 

